### PR TITLE
S3 plugin: don't use a default port

### LINKED ISSUE #247: S3 endpoint should be 's3.amazonaws.com' error
Running a backup job using `s3` as the store we got an error complaining about the s3 endpoint: `Amazon S3 endpoint should be s3.amazonaws.com`. 1st we tried with empty `s3_host`, later specifying `s3.amazonaws.com`, but both approaches failed.

Seems that at the minio connection, the plugin is [adding](https://github.com/starkandwayne/shield/blob/master/plugin/s3/plugin.go#L367) the `s3_port` (even if it is not specified as part of the config). Minio then [checks](https://github.com/starkandwayne/shield/blob/b9f27dd04474b01367124acd2f729712b3d5e475/vendor/github.com/starkandwayne/minio-go/utils.go#L215) if this is a valid AWS URL, but as the host contains the port, minio [does not recognize](https://github.com/starkandwayne/shield/blob/b9f27dd04474b01367124acd2f729712b3d5e475/vendor/github.com/starkandwayne/minio-go/utils.go#L169) it as a valid AWS endpoint.

Full log:

```
UUID:       2d0f6d45-9d52-49f1-a13e-ac76d5797413
Owner:      ubuntu@
Type:       backup
Status:     failed
            
Started at: Wed, 01 Feb 2017 10:28:39 +0000
Stopped at: Wed, 01 Feb 2017 10:28:39 +0000
            
Job:        mysql-backups (d0de68c0-735f-4c0a-b345-3f92afb24a9e)
            
Log:        Validating environment...
            =========================
            SHIELD_OP ... found
            SHIELD_STORE_PLUGIN ... found
            SHIELD_STORE_ENDPOINT ... found
            SHIELD_TARGET_PLUGIN ... found
            SHIELD_TARGET_ENDPOINT ... found
            OK
            
            Validating TARGET plugin `mysql`...
            ===================================
            DEBUG> 'validate' action requested with options plugin.PluginOpts{Debug:false, Version:false, Action:"validate", Info:struct {}{}, Validate:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:"{\"mysql_user\":\"root\",\"mysql_password\":\"<REDACTED>\",\"mysql_host\":\"<REDACTED>\",\"mysql_post\":\"3306\"}"}, Backup:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Restore:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Store:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing store endpoint'\"" }{Endpoint:""}, Retrieve:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing retrieve endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to retrieve from storage'\"" }{Endpoint:"", Key:""}, Purge:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing purge endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to purge from storage'\"" }{Endpoint:"", Key:""}}
            ✓ mysql_host          <REDACTED>
            ✓ mysql_port          using default port 3306
            ✓ mysql_user          root
            ✓ mysql_password      <REDACTED>
            ✓ mysql_read_replica  no read replica
            ✓ mysql_options  no options given - using defaults
            DEBUG> 'validate' action returned <nil>
            OK
            
            Validating STORE plugin `s3`...
            ===============================
            DEBUG> 'validate' action requested with options plugin.PluginOpts{Debug:false, Version:false, Action:"validate", Info:struct {}{}, Validate:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:"{\"access_key_id\":\"<REDACTED>\",\"bucket\":\"epc-poc\",\"prefix\":\"backups\",\"secret_access_key\":\"<REDACTED>\",\"signature_version\":\"4\",\"s3_host\":\"s3.amazonaws.com\"}"}, Backup:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Restore:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Store:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing store endpoint'\"" }{Endpoint:""}, Retrieve:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing retrieve endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to retrieve from storage'\"" }{Endpoint:"", Key:""}, Purge:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing purge endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to purge from storage'\"" }{Endpoint:"", Key:""}}
            ✓ s3_host              s3.amazonaws.com
            ✓ access_key_id        <REDACTED>
            ✓ s3_port               443
            ✓ secret_access_key    <REDACTED>
            ✓ bucket               epc-poc
            ✓ prefix               backups
            ✓ signature_version    4
            ✓ socks5_proxy         (no proxy will be used)
            ✓ skip_ssl_validation  no, SSL WILL be validated
            DEBUG> 'validate' action returned <nil>
            OK
            
            Running backup task (using bzip2 compression)
            =============================================
            DEBUG> 'backup' action requested with options plugin.PluginOpts{Debug:false, Version:false, Action:"backup", Info:struct {}{}, Validate:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Backup:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:"{\"mysql_user\":\"root\",\"mysql_password\":\"<REDACTED>\",\"mysql_host\":\"<REDACTED>\",\"mysql_post\":\"3306\"}"}, Restore:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Store:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing store endpoint'\"" }{Endpoint:""}, Retrieve:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing retrieve endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to retrieve from storage'\"" }{Endpoint:"", Key:""}, Purge:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing purge endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to purge from storage'\"" }{Endpoint:"", Key:""}}
            DEBUG> MYSQL_USER: 'root'
            DEBUG> MYSQL_PWD: '<REDACTED>'
            DEBUG> MYSQL_HOST: '<REDACTED>'
            DEBUG> MYSQL_PORT: '3306'
            DEBUG> MYSQL_READ_REPLICA: ''
            DEBUG> MYSQL_OPTIONS: ''
            DEBUG> MYSQL_DB: ''
            DEBUG> MYSQL_BIN_DIR: '/var/vcap/packages/shield-mysql/bin'
            DEBUG> Executing: `/var/vcap/packages/shield-mysql/bin/mysqldump  --all-databases -h <REDACTED> -P 3306 -u root`
            DEBUG> Executing '/var/vcap/packages/shield-mysql/bin/mysqldump' with arguments [--all-databases -h <REDACTED> -P 3306 -u root]
            DEBUG> 'store' action requested with options plugin.PluginOpts{Debug:false, Version:false, Action:"store", Info:struct {}{}, Validate:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Backup:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Restore:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing backup target'\"" }{Endpoint:""}, Store:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing store endpoint'\"" }{Endpoint:"{\"access_key_id\":\"<REDACTED>\",\"bucket\":\"epc-poc\",\"prefix\":\"backups\",\"secret_access_key\":\"<REDACTED>\",\"signature_version\":\"4\",\"s3_host\":\"s3.amazonaws.com\"}"}, Retrieve:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing retrieve endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to retrieve from storage'\"" }{Endpoint:"", Key:""}, Purge:struct { Endpoint string "goptions:\"-e, --endpoint, obligatory, description='JSON string representing purge endpoint'\""; Key string "goptions:\"-k, --key, obligatory, description='Key of blob to purge from storage'\"" }{Endpoint:"", Key:""}}
            DEBUG> 'store' action returned minio.ErrorResponse{XMLName:xml.Name{Space:"", Local:""}, Code:"InvalidArgument", Message:"Amazon S3 endpoint should be 's3.amazonaws.com'.", BucketName:"", Key:"", RequestID:"minio", HostID:"", AmzBucketRegion:""}
            Amazon S3 endpoint should be 's3.amazonaws.com'.
            DEBUG> 'backup' action returned <nil>
            
            EXITING 4
            TASK FAILED!!  shield worker 1 failed to execute the command against the remote agent 10.0.7.17:5444 (Process exited with: 4. Reason was:  ())
            TASK FAILED!! No restore key detected in worker 1. Cowardly refusing to create an archive record

```

### PR DESCRIPTION
**Pull request by frodenas**
Fixes #247

### CODE REVIEW

**Comment by jhunt** [comment]
Should we validate/document that if `s3_port` is given, `s3_host` is also non-empty?

**Comment by frodenas** [comment]
I don't have a strong opinion, I'm fine not validating this, but up to you.

File: plugin/s3/plugin.go [hunk 1]
```diff
@@ -27,7 +27,7 @@
 //        "prefix":              "/path/inside/bucket/to/place/backup/data",
 //        "signature_version":   "4",  # should be 2 or 4. Defaults to 4
 //        "socks5_proxy":        ""    # optionally defined SOCKS5 proxy to use for the s3 communications
-//        "s3_port":             "443"
+//        "s3_port":             ""    # optionally defined port to use for the s3 communications
```
Comment (jhunt) [inline plugin/s3/plugin.go]: Please document that s3_port now requires s3_host to function properly.

File: plugin/s3/plugin.go [hunk 2]
```diff
@@ -142,12 +141,10 @@ func (p S3Plugin) Validate(endpoint plugin.ShieldEndpoint) error {
 		ansi.Printf("@G{\u2713 access_key_id}        @C{%s}\n", s)
 	}
 
-	s, err = endpoint.StringValueDefault("s3_port","443")
+	s, err = endpoint.StringValueDefault("s3_port", "")
```
Comment (jhunt) [inline plugin/s3/plugin.go]: Please validate (somewhere in here) that s3_host is specified if s3_port is given.

**Comment by frodenas** [comment]
Added a comment and validation.

**Pull request merged**